discovery/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,220 @@
+		log.Errorf("Failed to query %v: %s", backendName, err)
+		return false
+	}
+	didChange := zookeeperCompareForChange(zookeeperUpstreams[backendName], services)
+	if didChange || len(services) == 0 {
+		// We don't want to cause an onChange event the first time we read-in
+		// but we do want to make sure we've written the key for this map
+		zookeeperUpstreams[backendName] = services
+	}
+	return didChange
+}
+
+func (c *ZooKeeper) getServices(appName string) ([]ServiceNode, error) {
+	var services []ServiceNode
+
+	key := c.getAppKey(appName)
+	children, _, error := c.Connection.Children(key)
+	if error != nil {
+		return services, error
+	}
+	for i := range children {
+		path := fmt.Sprintf("%s/%s", key, children[i])
+		data, _, error := c.Connection.Get(path)
+		if error != nil {
+			return services, error
+		}
+		srv, error := decodeZooKeeperNodeValue(data)
+		if error != nil {
+			return services, error
+		}
+		services = append(services, srv)
+	}
+	return services, nil
+}
+
+// Compare the two arrays to see if the address or port has changed
+// or if we've added or removed entries.
+func zookeeperCompareForChange(existing, new []ServiceNode) (changed bool) {
+	if len(existing) != len(new) {
+		return true
+	}
+
+	sort.Sort(ByZooKeeperServiceID(existing))
+	sort.Sort(ByZooKeeperServiceID(new))
+	for i, ex := range existing {
+		if ex.Address != new[i].Address ||
+			ex.Port != new[i].Port {
+			return true
+		}
+	}
+	return false
+}
+
+func (c ZooKeeper) createParentPath(path string) error {
+	pathElements := strings.Split(path, "/")[1:]
+	sep := "/"
+	newPath := ""
+	for i := range pathElements {
+		newPath = strings.Join([]string{newPath, sep, pathElements[i]}, "")
+		if exists, _, _ := c.Connection.Exists(newPath); !exists {
+			if _, err := c.Connection.Create(newPath, nil, 0, zk.WorldACL(zk.PermAll)); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c ZooKeeper) registerService(service *discovery.ServiceDefinition) error {
+	k := c.getNodeKey(service)
+	if err := c.createParentPath(c.parentPath(service)); err != nil {
+		return err
+	}
+	value := encodeZooKeeperNodeValue(service)
+	if exists, _, _ := c.Connection.Exists(k); !exists {
+		if _, err := c.Connection.Create(k, []byte(value), 0, zk.WorldACL(zk.PermAll)); err != nil {
+			return err
+		}
+		_, _, ch, _ := c.Connection.GetW(k)
+		go func() {
+			select {
+			case ev := <-ch:
+				_, _, ch, _ = c.Connection.GetW(ev.Path)
+			case <-time.After(time.Duration(service.TTL) * time.Second):
+				log.Warningf("TTL expired, deregistering %s", k)
+				c.Deregister(service)
+			}
+		}()
+	}
+	return nil
+}
+
+func encodeZooKeeperNodeValue(service *discovery.ServiceDefinition) string {
+	node := &ServiceNode{
+		ID:      service.ID,
+		Name:    service.Name,
+		Address: service.IPAddress,
+		Port:    service.Port,
+	}
+	json, err := json.Marshal(&node)
+	if err != nil {
+		log.Warnf("Unable to encode service: %s", err)
+		return ""
+	}
+	return string(json)
```

Comment:
You're coercing the `[]byte` to a string but then coercing it back to `[]byte` when you use it in `c.Connection.Create`, etc. You can just have this function return `[]byte` and be good-to-go.